x/stakeibc/keeper/msg_server_lsm_liquid_stake.go

```diff
@@ -177,32 +130,59 @@ func (k Keeper) SubmitValidatorSlashQuery(ctx sdk.Context, lsmLiquidStake types.
 	return nil
 }
 
-// FinishLSMLiquidStake finishes the liquid staking flow by sending a user an stToken and
-// IBC transfering the LSM Token to the host zone
+// FinishLSMLiquidStake finishes the liquid staking flow by escrowing the LSM token,
+//   sending a user their stToken, and then IBC transfering the LSM Token to the host zone
 //
-// If the validator exchange rate query interrupted the transaction, this function is called
-// asynchronously after the query callback
-// If no validator exchange rate query was needed, this is called synchronously after StartLSMLiquidStake
+// If the slash query interrupted the transaction, this function is called
+//   asynchronously after the query callback
+// If no slash query was needed, this is called synchronously after StartLSMLiquidStake
 func (k Keeper) FinishLSMLiquidStake(ctx sdk.Context, lsmLiquidStake types.LSMLiquidStake) error {
+	hostZone := lsmLiquidStake.HostZone
+	lsmTokenDeposit := lsmLiquidStake.Deposit
+
+	// Validate the LSM liquid stake message parameters again, in the event that the transaction
+	// was interrupted by the slash query.
+	// The most significant check here is that the user still has sufficient balance for this LSM liquid stake
+	lsmLiquidStakeMsg := types.MsgLSMLiquidStake{
+		Creator:          lsmTokenDeposit.StakerAddress,
+		LsmTokenIbcDenom: lsmTokenDeposit.IbcDenom,
+		Amount:           lsmTokenDeposit.Amount,
+	}
+	if _, err := k.ValidateLSMLiquidStake(ctx, lsmLiquidStakeMsg); err != nil {
+		return err
+	}
+
+	// Get the staker's address and the host zone's deposit account address (which will custody the tokens)
+	liquidStakerAddress := sdk.MustAccAddressFromBech32(lsmTokenDeposit.StakerAddress)
+	hostZoneAddress, err := sdk.AccAddressFromBech32(hostZone.DepositAddress)
+	if err != nil {
+		return errorsmod.Wrapf(err, "host zone address is invalid")
+	}
+
+	// Transfer the LSM token to the deposit account
+	lsmIBCToken := sdk.NewCoin(lsmTokenDeposit.IbcDenom, lsmTokenDeposit.Amount)
```

Comment:
So we don't actually take control of the LSM token from the user and hand them the stTokens until the finish part of the function because we may have had to wait for the query to give us an answer back and now this avoids any need to refund, but in the start part of the function we set the record for the LSMTokenDeposit... shouldn't we also wait until the transaction is actually going to take place to set this record?

Otherwise we need to explicitly delete that record if anything has gone wrong enough to prevent the user from getting their stTokens but I don't see anywhere that happens in the finish part of the function unless by throwing an error in the finish part it is part of the same state updates as in the start part so throwing an error would negate any changes which would have happened to the records?

Reply:
sorry somehow just seeing this! Very good point, I think we should delete the LSMTokenDeposit if the finish function fails. We could also keep it for monitoring purposes, but since it's easy for the user to manipulate, I'd vote removing it!

Reply:
Going to handle this in the other PR since it's part of the callback